hedera-node/hedera-networkadmin-service-impl/src/main/java/com/hedera/node/app/service/networkadmin/impl/handlers/FreezeHandler.java

```diff
@@ -72,51 +109,90 @@ public void preHandle(@NonNull final PreHandleContext context) throws PreCheckEx
                 // reject any freeze transactions that do not set freezeType or set it to UNKNOWN_FREEZE_TYPE
             case UNKNOWN_FREEZE_TYPE -> throw new PreCheckException(INVALID_FREEZE_TRANSACTION_BODY);
 
                 // FREEZE_ONLY requires a valid start_time
-            case FREEZE_ONLY -> verifyFreezeStartTimeIsInFuture(freezeTxn, txnId.transactionValidStart());
-
-                // PREPARE_UPGRADE requires valid update_file and file_hash values
-            case PREPARE_UPGRADE -> verifyUpdateFileAndHash(freezeTxn, specialFileStore);
-
                 // FREEZE_UPGRADE and TELEMETRY_UPGRADE require a valid start_time and valid update_file and
                 // file_hash values
-            case FREEZE_UPGRADE, TELEMETRY_UPGRADE -> {
-                verifyFreezeStartTimeIsInFuture(freezeTxn, txnId.transactionValidStart());
-
-                // from proto specs, it looks like update file not required for FREEZE_UPGRADE and TELEMETRY_UPGRADE
-                // but specs aren't very clear
-                // current code in FreezeTransitionLogic checks for the file in specialFiles
-                // so we will do the same
-                verifyUpdateFileAndHash(freezeTxn, specialFileStore);
-            }
+            case FREEZE_ONLY, FREEZE_UPGRADE, TELEMETRY_UPGRADE -> verifyFreezeStartTimeIsInFuture(
+                    freezeTxn, txValidStart);
 
-                // FREEZE_ABORT does not require any additional checks
-            case FREEZE_ABORT -> {
+                // PREPARE_UPGRADE, FREEZE_ABORT do not require any additional checks
+            default -> {
                 // do nothing
             }
         }
+    }
 
-        // no need to add any keys to the context because this transaction does not require any signatures
-        // it must be submitted by an account with superuser privileges, that is checked during ingest
+    public void handle(
+            @NonNull final TransactionBody txn,
+            @NonNull final NetworkAdminServiceConfig adminServiceConfig,
+            @NonNull final ReadableSpecialFileStore specialFileStore,
+            @NonNull final WritableFreezeStore freezeStore) {
+        requireNonNull(txn);
+        final FreezeTransactionBody freezeTxn = txn.freezeOrThrow();
+        final FileID updateFileNum =
+                freezeTxn.updateFile(); // only some freeze types require this, it may be null for others
+
+        validateSemantics(freezeTxn, specialFileStore, updateFileNum);
+
+        final FreezeUpgradeActions upgradeActions = new FreezeUpgradeActions(adminServiceConfig, freezeStore);
+        final Timestamp freezeStartTime = freezeTxn.startTime(); // may be null for some freeze types
+        final Instant freezeStartTimeInstant = freezeStartTime == null
+                ? null
+                : Instant.ofEpochSecond(freezeStartTime.seconds(), freezeStartTime.nanos());
+
+        switch (freezeTxn.freezeType()) {
```

Comment:
You should probably assign the result of this switch expression to a variable, and possibly join the CompletableFuture, since Handle is supposed to complete all work before returning (or is freeze a special case that does not need to be deterministic?).

Reply:
By "assign the result of this switch expression to a variable" do you mean
```
FreezeType freezeType = freezeTxn.FreezeType();
switch(freezeType) {
    //etc
```
? What is the reason for doing that?

Reply:
No need to join the CompletableFuture. There are typically hours to days after the upgrade file is unzipped before the update is triggered. We know that it is complete because a marker file is written. We don't want to hold up the consensus round waiting for it to complete.

Reply:
Assigning the result:
`var switchResult = switch(freezetype {`
Join:
  The problem is that the unzip will never happen if you don't join the result.  The default fork-join pool (which is what you're using) does not run tasks that are never joined in most cases (it's not entirely deterministic, but generally true).
You can pass in your own fork-join pool to the `runAsync` method that has the `asyncMode` attribute set `true` if you really do want it to run "eventually" and do not want to manage the Future.
If you do setup a fork-join pool, it may be best to do so in the class constructor so it can possibly be injected (or a PoolManager injected) in the future.